pkg/services/notifications/smtp_test.go

```diff
@@ -0,0 +1,108 @@
+package notifications
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBuildMail(t *testing.T) {
+	cfg := setting.NewCfg()
+	cfg.Smtp.ContentTypes = []string{"text/html", "text/plain"}
+
+	sc, err := NewSmtpClient(cfg.Smtp)
+	require.NoError(t, err)
+
+	message := &Message{
+		To:      []string{"to@address.com"},
+		From:    "from@address.com",
+		Subject: "Some subject",
+		Body: map[string]string{
+			"text/html":  "Some HTML body",
+			"text/plain": "Some plain text body",
+		},
+		ReplyTo: []string{"from@address.com"},
+	}
+
+	t.Run("When building email", func(t *testing.T) {
+		email := sc.buildEmail(message)
+
+		buf := new(bytes.Buffer)
+		_, err := email.WriteTo(buf)
+		require.NoError(t, err)
+
+		assert.Contains(t, buf.String(), "Some HTML body")
+		assert.Contains(t, buf.String(), "Some plain text body")
+		assert.Less(t, strings.Index(buf.String(), "Some plain text body"), strings.Index(buf.String(), "Some HTML body"))
+	})
+}
+
+func TestSmtpDialer(t *testing.T) {
+	t.Run("When SMTP hostname is invalid", func(t *testing.T) {
+		cfg := createSmtpConfig()
+		cfg.Smtp.Host = "invalid%hostname:123:456"
+		client, err := ProvideSmtpService(cfg)
+		require.NoError(t, err)
+		message := &Message{
+			To:          []string{"asdf@grafana.com"},
+			SingleEmail: true,
+			Subject:     "subject",
+			Body: map[string]string{
+				"text/html":  "body",
+				"text/plain": "body",
+			},
+		}
+
+		count, err := client.Send(message)
+
+		require.Equal(t, 0, count)
+		require.Error(t, err)
+	})
+
+	t.Run("When SMTP port is invalid", func(t *testing.T) {
+		cfg := createSmtpConfig()
+		cfg.Smtp.Host = "invalid%hostname:123a"
+		client, err := ProvideSmtpService(cfg)
+		require.NoError(t, err)
+		message := &Message{
+			To:          []string{"asdf@grafana.com"},
+			SingleEmail: true,
+			Subject:     "subject",
+			Body: map[string]string{
+				"text/html":  "body",
+				"text/plain": "body",
+			},
+		}
+
+		count, err := client.Send(message)
+
+		require.Equal(t, 0, count)
+		require.Error(t, err)
```

Comment:
After looking through the `testify.require` documentation, I didn't find a very clean way to assert on the content of an error if the error is untyped. These tests aren't perfect because they only assert that _some_ error occurred, not that the expected error was surfaced, which could lead to them being ineffective if the underlying code changes enough.

Is there a cleaner way of asserting on the actual string content of an error that doesn't require lots of awkward manipulation in the test logic?

Reply:
`require.EqualError`?